robot_sort/robot_sort.py

```diff
@@ -1,112 +1,188 @@
-        """
-        self._light = "OFF"
-    def light_is_on(self):
-        """
-        Returns True if the robot's light is on and False otherwise.
-        """
-        return self._light == "ON"
-
-    def sort(self):
-        """
-        Sort the robot's list.
-        """
-        # Fill this out
-        pass
+	def __init__(self, l):
+		"""
+		SortingRobot takes a list and sorts it.
+		"""
+		self._list = l          # The list the robot is tasked with sorting
+		self._item = None       # The item the robot is holding
+		self._position = 0      # The list position the robot is at
+		self._light = "OFF"     # The state of the robot's light
+		self._time = 0          # A time counter (stretch)
+
+	def can_move_right(self):
+		"""
+		Returns True if the robot can move right or False if it's
+		at the end of the list.
+		"""
+		return self._position < len(self._list) - 1
+
+	def can_move_left(self):
+		"""
+		Returns True if the robot can move left or False if it's
+		at the start of the list.
+		"""
+		return self._position > 0
+
+	def move_right(self):
+		"""
+		If the robot can move to the right, it moves to the right and
+		returns True. Otherwise, it stays in place and returns False.
+		This will increment the time counter by 1.
+		"""
+		self._time += 1
+		if self._position < len(self._list) - 1:
+			self._position += 1
+			return True
+		else:
+			return False
+
+	def move_left(self):
+		"""
+		If the robot can move to the left, it moves to the left and
+		returns True. Otherwise, it stays in place and returns False.
+		This will increment the time counter by 1.
+		"""
+		self._time += 1
+		if self._position > 0:
+			self._position -= 1
+			return True
+		else:
+			return False
+
+	def swap_item(self):
+		"""
+		The robot swaps its currently held item with the list item in front
+		of it.
+		This will increment the time counter by 1.
+		"""
+		self._time += 1
+		# Swap the held item with the list item at the robot's position
+		self._item, self._list[self._position] = self._list[self._position], self._item
+
+	def compare_item(self):
+		"""
+		Compare the held item with the item in front of the robot:
+		If the held item's value is greater, return 1.
+		If the held item's value is less, return -1.
+		If the held item's value is equal, return 0.
+		If either item is None, return None.
+		"""
+		if self._item is None or self._list[self._position] is None:
+			return None
+		elif self._item > self._list[self._position]:
+			return 1
+		elif self._item < self._list[self._position]:
+			return -1
+		else:
+			return 0
+
+	def set_light_on(self):
+		"""
+		Turn on the robot's light
+		"""
+		self._light = "ON"
+
+	def set_light_off(self):
+		"""
+		Turn off the robot's light
+		"""
+		self._light = "OFF"
+
+	def light_is_on(self):
+		"""
+		Returns True if the robot's light is on and False otherwise.
+		"""
+		return self._light == "ON"
+
+	def sort(self):
+		"""
+		Sort the robot's list.
+		"""
+		#swap_item
```

Comment:
@Trosaar , The few things you could have mentioned in the pseudocode for your robot's sorting algorithm:

(a) **Line 141.50** could have been: Having swapped the "**None**" object for the item in the first position, the robot is now at the second position and is now beginning a "comparison protocol" where it is comparing the value of **the item in its possession** to the value of **the item in front of it** (_the item in the second position_).

(b) **Line 142.50** could have been: Two things can happen during this "comparison protocol". First, if the value of **the item in the robot's possession** is **SMALLER** than the value of **the item in front of it** (in the second position), those two items are considered to be "_already sorted_" so the robot will **NOT** swap those two items, but instead move back to left one position and swap the item in its possession (the item that was _originally_ in the first position) with the "None" object (that is currently at the first position). Then, the robot will take the "**None**" object and move forward to the right one space, **back** to the second position. Here, the robot will swap it's "**None**" object for the item in front of it (in the second position) and will move forward to the right one space to repeat the "comparison procedure" as defined above.

(c) **Line 142.75** could have been: The second thing that could happen during this "comparison protocol" would occur if the value of **the item in the robot's possession** is **GREATER** than the value of **the item in front of it** (in the second position). Then, those two items being compared would be considered  "**unsorted**" and so the robot **WILL**, indeed, swap those two items, and **THEN** move back to left one position. Now, standing in front of the "**None**" object again, the robot will swap the **smaller, recently-swapped item in its possession** with the "**None**" object (that is currently at the first position). Then, the robot will take the "**None**" object and move forward to the right one space. Here, the robot will swap it's "**None**" object for the item in front of it and then it will move forward to the right one space to repeat this "comparison procedure" as defined above again, until the robot does no swaps during a complete pass through of all the items.

This is another use case for the light on top of the robot's head. Some people used the light as a way to start the sorting process (i.e. - this is where we enter the "while" loop). Then, they would shut off the light as soon as they entered the "while" loop and THEN use the light as an indicator of whether or not AT LEAST one swap occurred during an entire pass through of all the items! If the robot's light turned on EVEN once during a pass through of all those items, the robot knew that everything was not fully sorted. However, when the robot's light did NOT turn back on after it had initially turned off upon entering the "while" loop, during an entire pass through of all the items, the robot would understand that all of the items were sorted because there were no swaps that occurred.

Hope that makes sense, @Trosaar  ! 😅 👍